### ISSUE TITLE
Ad is detected but the ad continues to play.

### ISSUE THREAD

**Issue by debasish-patnaik** [abertschi/ad-free#41]
I am using a OnePlus 7T, ad-free detects that an ad is being played but it does nothing about it, the ad continues to play.

PS: I'm using the mute plug-in.

**Comment by abertschi** [comment]
sounds like a device specific issue that requires debugging. as I don't own your device, I cannot investigate it.

**Comment by abertschi** [comment]
Does it happen all the time? What happens if you play music, open ad-free, navigate to the mute-plugin screen and test the plugin (tab onto the music note icon at the top right). Is the music muted?

**Comment by debasish-patnaik** [comment]
I was using a bluetooth earphone and it wasn't working at the time. After that I tried with my wired earphone and it worked and now when I reconnected my bluetooth earphone it seems to work. Don't know why it wasn't working the first time though.

**Comment by debasish-patnaik** [comment]
The issue still happens while using my bluetooth headset, disconnecting the headset and reconnecting it resolves the issue but it displays the same behaviour after some time.

**Comment by bertolys** [comment]
Hi Andrin, since I updated from Android 9 to 10, I have the same behavior. I did not try with cable tough, only Bluetooth, and the ads are just playing. Taping on the music note doesn't mute the music. I have an OnePlus 6T. Upgraded, uninstalled the app did nothing. Thanks for looking into this.

**Comment by abertschi** [comment]
I haven't tested the app on Android Q, as issues related to mute cannot be easily tested/debugged in a emulator and my devices are below Q. The behavior you described seems to be device specific as I do not have these issues on my devices. So unless someone debugs the issue on your device, it is likely that it will not get fixed.

**Comment by abertschi** [comment]
> Hi Andrin, since I updated from Android 9 to 10, I have the same behavior. I did not try with cable tough, only Bluetooth, and the ads are just playing. Taping on the music note doesn't mute the music. I have an OnePlus 6T. Upgraded, uninstalled the app did nothing. Thanks for looking into this.

Sometimes ad-free's background service is killed by the OS to save battery. In this case re-enabling the notification listener in the settings can fix undetected ads. Also remove ad-free from the list of apps that run in battery saved mode can fix issues. Issues for which ad-free detects ads but does not mute need debugging to figure out what's wrong.

**Comment by debasish-patnaik** [comment]
Hello Andrin, upon tinkering around with your code (I've got only basic understanding of android app development) I found out that when I changed the method you used to set volume from adjustStreamVolume to setStreamVolume it worked and while unmuting using the same method the volume was set to max, so I set the volume to volume it was before using the musicStreamVolume variable that stored the previous volume.

From this:
`fun muteMusicStream() {
        debug("current volume " + musicStreamVolume)
        info("muting audio")

        if (musicStreamIsMuted) {
            return
        }
        musicStreamIsMuted = true
        val am = context.getSystemService(Context.AUDIO_SERVICE) as AudioManager
        musicStreamVolume = am.getStreamVolume(AudioManager.STREAM_MUSIC)


        am.adjustStreamVolume(AudioManager.STREAM_MUSIC, AudioManager.ADJUST_MUTE, 0)

    }


    fun unmuteMusicStream() {
        info("Unmuting audio....")
        if (!musicStreamIsMuted) {
            return
        }
        musicStreamIsMuted = false
        val am = context.getSystemService(Context.AUDIO_SERVICE) as AudioManager
        am.adjustStreamVolume(AudioManager.STREAM_MUSIC, AudioManager.ADJUST_UNMUTE, 0)
    }` 
to this:
`fun muteMusicStream() {
        debug("current volume " + musicStreamVolume)
        info("muting audio")

        if (musicStreamIsMuted) {
            return
        }
        musicStreamIsMuted = true
        val am = context.getSystemService(Context.AUDIO_SERVICE) as AudioManager
        musicStreamVolume = am.getStreamVolume(AudioManager.STREAM_MUSIC)


        am.setStreamVolume(AudioManager.STREAM_MUSIC, AudioManager.ADJUST_MUTE, 0)

    }


    fun unmuteMusicStream() {
        info("Unmuting audio....")
        if (!musicStreamIsMuted) {
            return
        }
        musicStreamIsMuted = false
        val am = context.getSystemService(Context.AUDIO_SERVICE) as AudioManager
        am.setStreamVolume(AudioManager.STREAM_MUSIC, musicStreamVolume, 0)
    }`

**Comment by debasish-patnaik** [comment]
Hello Andrin, upon tinkering around with your code (I've got only basic understanding of android app development) I found out that when I changed the method you used to set volume from adjustStreamVolume to setStreamVolume it worked and while unmuting using the same method the volume was set to max, so I set the volume to volume it was before using the musicStreamVolume variable that stored the previous volume.

Original:

`fun muteMusicStream() {
        debug("current volume " + musicStreamVolume)
        info("muting audio")

        if (musicStreamIsMuted) {
            return
        }
        musicStreamIsMuted = true
        val am = context.getSystemService(Context.AUDIO_SERVICE) as AudioManager
        musicStreamVolume = am.getStreamVolume(AudioManager.STREAM_MUSIC)


        am.adjustStreamVolume(AudioManager.STREAM_MUSIC, AudioManager.ADJUST_MUTE, 0)

    }


    fun unmuteMusicStream() {
        info("Unmuting audio....")
        if (!musicStreamIsMuted) {
            return
        }
        musicStreamIsMuted = false
        val am = context.getSystemService(Context.AUDIO_SERVICE) as AudioManager
        am.adjustStreamVolume(AudioManager.STREAM_MUSIC, AudioManager.ADJUST_UNMUTE, 0)
    }`

Modified:

`fun muteMusicStream() {
        debug("current volume " + musicStreamVolume)
        info("muting audio")

        if (musicStreamIsMuted) {
            return
        }
        musicStreamIsMuted = true
        val am = context.getSystemService(Context.AUDIO_SERVICE) as AudioManager
        musicStreamVolume = am.getStreamVolume(AudioManager.STREAM_MUSIC)


        am.setStreamVolume(AudioManager.STREAM_MUSIC, AudioManager.ADJUST_MUTE, 0)

    }


    fun unmuteMusicStream() {
        info("Unmuting audio....")
        if (!musicStreamIsMuted) {
            return
        }
        musicStreamIsMuted = false
        val am = context.getSystemService(Context.AUDIO_SERVICE) as AudioManager
        am.setStreamVolume(AudioManager.STREAM_MUSIC, musicStreamVolume, 0)
    }`

**Comment by debasish-patnaik** [comment]
Hello Andrin, upon tinkering around with your code (I've got only basic understanding of android app development) I changed the method you used to set volume from adjustStreamVolume to setStreamVolume and it worked. Upon unmuting, the volume kept resetting to max volume, so I set the volume to it's previous value before muting and it seems to be working as expected.

PS: I don't know why this fix works. Also, thanks for this amazing app.

**Comment by abertschi** [comment]
:+1: , will look into it. Probably on the weekend.

**Comment by bertolys** [comment]
> 
> 
> > Hi Andrin, since I updated from Android 9 to 10, I have the same behavior. I did not try with cable tough, only Bluetooth, and the ads are just playing. Taping on the music note doesn't mute the music. I have an OnePlus 6T. Upgraded, uninstalled the app did nothing. Thanks for looking into this.
> 
> Sometimes ad-free's background service is killed by the OS to save battery. In this case re-enabling the notification listener in the settings can fix undetected ads. Also remove ad-free from the list of apps that run in battery saved mode can fix issues. Issues for which ad-free detects ads but does not mute need debugging to figure out what's wrong.

Hi Ardin, thanks, I will try these and let you know.

**Comment by debasish-patnaik** [comment]
Hey Andrin, looking into the issue further, I found out that bluetooth absolute volume was causing the issue. Basically, when adjustStreamVolume method is used to mute the volume, my bluetooth earphones resets the volume back to what the device remembers as the last setting.

Upon disabling bluetooth absolute volume from developer options, this issue was fixed even with adjustStreamVolume. But this has the effect of requiring a user to control two (headphone and phone) volumes separately.

However as I have said above, if setStreamVolume is used in place of adjustStreamVolume it works as expected even if bluetooth absolute volume is turned on.
